<h1> Data Analytics Project </h1>

<h3> United Kingdom Road Accident Data Analysis </h3>
<h3> Inclusive Years: 2019 - 2022 </h3>
<p> Analyst: Nicole Patrice Torrero <p>

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

<h1> Data Preparation </h1>

In [3]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1> Converting Object to DateTime Data Type </h1>

In [4]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [5]:
accident.dtypes
# accident['Accident Date']

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

<h1> Data Cleansing </h1>
<h3> Filling out null Values</h3>

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown surface condition')  
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna('unaccounted')

<h1> Categorical Data Fields </h1>

In [7]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['Accident_Severity'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')


<h1> Extracting Date information using pandas date time </h1>

In [8]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['Day of the Week'] = accident['Accident Date'].dt.dayofweek

In [9]:
df_2019 = accident[accident['Year'] == 2019]

<h1> Univariate Analysis from year 2019 </h1>
<h3> Question 1. How many incidents happened under light conditions? </h3>

In [10]:
df_2019['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

<h1> Insight 1 </h1>
<p> Daylight constitute to 132% of the total accidents from year 2019 </p>

<h1> Question. 2 </h1>
<h3> Which way shows a high number of slight accidents? </h3>

In [11]:
g = accident.groupby(['Accident_Severity', 'Weather_Conditions', 'Road_Type']).size()
g


Accident_Severity  Weather_Conditions  Road_Type         
Fatal              Fine + high winds   Dual carriageway        32
                                       One way street           5
                                       Roundabout               5
                                       Single carriageway     133
                                       Slip road                0
                                                             ... 
Slight             unaccounted         One way street         298
                                       Roundabout            1161
                                       Single carriageway    8755
                                       Slip road              142
                                       unaccounted            831
Length: 162, dtype: int64

<h1> Insight 2 </h1>
<p> One-way streets and roundabouts show a high number of slight accidents. </p>

<h1> Question 3. </h1>
<h3>  What are the most common accidents in urban or rural areas?</h3>


In [12]:
b = accident.groupby(['Accident_Severity', 'Urban_or_Rural_Area']).size()
b

Accident_Severity  Urban_or_Rural_Area
Fatal              Rural                    5601
                   Unallocated                 0
                   Urban                    3060
                   unaccounted                 0
Serious            Rural                   37312
                   Unallocated                 1
                   Urban                   50902
                   unaccounted                 2
Slight             Rural                  196077
                   Unallocated                10
                   Urban                  367701
                   unaccounted                13
dtype: int64

<h1> Insight 3</h1>
<p> Fatal accidents are more common in rural areas a total of 5,601, compared to urban areas which is 3,060. </p>

<h1> Insight 4 </h1>
<h3> Did serious accidents increase or decrease over time? </h3>

In [13]:
accident['Accident Year'] = accident['Accident Date'].dt.year
a = accident.groupby(['Accident Year', 'Accident_Severity'])['Number_of_Casualties'].sum()
a

Accident Year  Accident_Severity
2019           Fatal                  5351
               Serious               36096
               Slight               206333
2020           Fatal                  4398
               Serious               33987
               Slight               192520
2021           Fatal                  3879
               Serious               32311
               Slight               185956
2022           Fatal                  2855
               Serious               27045
               Slight               165837
Name: Number_of_Casualties, dtype: int64

<h1> Insight 4 </h1>
<p>Serious accidents decreased from 36,096 in 2019 to 27,045 in 2022, showing a clear reduction over the years.</p>

<h1> Question 5. </h1>
<h3> Do most fatal accidents happen during the day or at night, and which is more dangerous? </h3>

In [14]:
b = accident.groupby(['Accident_Severity', 'Light_Conditions']).size()
b

Accident_Severity  Light_Conditions           
Fatal              Darkness - lighting unknown        68
                   Darkness - lights lit            1860
                   Darkness - lights unlit            45
                   Darkness - no lighting           1612
                   Daylight                         5076
Serious            Darkness - lighting unknown       794
                   Darkness - lights lit           19130
                   Darkness - lights unlit           360
                   Darkness - no lighting           7174
                   Daylight                        60759
Slight             Darkness - lighting unknown      5622
                   Darkness - lights lit          108345
                   Darkness - lights unlit          2138
                   Darkness - no lighting          28651
                   Daylight                       419045
dtype: int64

<h1> Insight 5 </h1>
<p> Most fatal accidents occur during daylight, but nighttime presents a higher risk. </p>

<h1> Question 6.</h1>
<h3> Which district areas have the highest accident frequency for each level of accident severity? </h3>

In [15]:
accidents_by_area_severity = accident.groupby(['District Area', 'Accident_Severity']).size().reset_index(name='Accident_Count')
accidents_by_area_severity

District Area Accident_Severity  Accident_Count
0         Fatal             Fatal            8661
1         Fatal           Serious               0
2         Fatal            Slight               0
3       Serious             Fatal               0
4       Serious           Serious           88217
5       Serious            Slight               0
6        Slight             Fatal               0
7        Slight           Serious               0
8        Slight            Slight          563801

<h1> Insight 6 </h1>
<p> The highest number of serious accidents occurred in a district, with a total of 88,217 cases, making it the area with the most severe incidents. </p>

<h1> Insight 7 </h1>
<h3> How many accidents occurred in each district for each level of severity? </h3>

In [16]:
urban_rural_accidents = accidents_by_area_severity.groupby(['District Area', 'Accident_Severity'])['Accident_Count'].sum().unstack().fillna(0)
urban_rural_accidents

Accident_Severity  Fatal  Serious  Slight
District Area                            
Fatal               8661        0       0
Serious                0    88217       0
Slight                 0        0  563801

<h1> Insight 7 </h1>
<ul>
    <li> Fatal Districts: Recorded 8,661 fatal accidents, but no serious or slight accidents. </li>
    <li> Serious Districts: Had 88,217 serious accidents, with no fatal or slight accidents. </li>
    <li> Slight Districts: Experienced 563,801 slight accidents, without any fatal or serious cases. </li>
</ul>

<h1> Question 8.</h1>
<h3> How many accidents of each severity (Fatal, Serious, Slight) were reported in the district area, and what is the breakdown of these accidents? </h3>

In [17]:
accidents_by_area = accident.groupby(['District Area', 'Accident_Severity']).size().reset_index(name='Accident_Count')
accidents_by_area


District Area Accident_Severity  Accident_Count
0         Fatal             Fatal            8661
1         Fatal           Serious               0
2         Fatal            Slight               0
3       Serious             Fatal               0
4       Serious           Serious           88217
5       Serious            Slight               0
6        Slight             Fatal               0
7        Slight           Serious               0
8        Slight            Slight          563801

<h1> Insight 8 </h1>
<p> Fatal accidents were significantly lower in number, with only 8,661 recorded, all of which were categorized as fatal. </p>

<h1> Question 9. </h1>
<h3> Does fog contribute to a high number of accidents on single carriageways? </h3>

In [18]:
weather_road_analysis = accident.groupby(['Weather_Conditions', 'Road_Type'])['Accident_Severity'].count().unstack()
weather_road_analysis


Road_Type              Dual carriageway  One way street  Roundabout  \
Weather_Conditions                                                    
Fine + high winds                  1519             158         540   
Fine no high winds                76916           11057       34667   
Fog or mist                         682              29         222   
Other                              2264             309        1109   
Raining + high winds               2033             153         563   
Raining no high winds             13044            1429        5347   
Snowing + high winds                185              14          30   
Snowing no high winds              1020              78         290   
unaccounted                        1761             332        1224   

Road_Type              Single carriageway  Slip road  unaccounted  
Weather_Conditions                                                 
Fine + high winds                    6178        109           50  
Fine no high winds                 389830       5520         2895  
Fog or mist                          2541         33           21  
Other                               13156        180          132  
Raining + high winds                 6703        107           56  
Raining no high winds               58581        865          430  
Snowing + high winds                  642         10            4  
Snowing no high winds                4751         69           30  
unaccounted                          9761        148          902

<h1> Insight 9 </h1>
<p> Fog causes fewer accidents with an overall 2,541 on single carriageways </p>

<h1> Insight 10 </h1>
<h3> Which day of the week has the highest number of accidents? </h3>

In [19]:
accident['Day_of_Week'] = pd.to_datetime(accident['Accident Date']).dt.day_name()
d = accident.groupby('Day_of_Week').size().sort_values(ascending=False)
d

Day_of_Week
Saturday     107178
Wednesday     99558
Thursday      99511
Friday        97900
Tuesday       94550
Sunday        89302
Monday        72680
dtype: int64

<h1> Insight 10 </h1>
<p>  Saturday has the highest number of accidents </p>

<h1> Insight 11 </h1>
<h3> In which month do fatal accidents occur the most and the least? </h3>

In [20]:
accident['Month'] = pd.to_datetime(accident['Accident Date']).dt.month
e = accident.groupby('Month')['Accident_Severity'].value_counts(normalize=True).unstack() * 100
np.round(e, 2)


Accident_Severity  Fatal  Serious  Slight
Month                                    
1                   1.40    12.64   85.96
2                   1.25    12.62   86.13
3                   1.26    13.16   85.58
4                   1.35    14.18   84.48
5                   1.35    13.70   84.95
6                   1.19    14.02   84.79
7                   1.31    13.69   85.00
8                   1.49    13.97   84.54
9                   1.28    14.15   84.57
10                  1.34    13.32   85.34
11                  1.24    12.52   86.24
12                  1.29    12.17   86.53

<h1> Insight 11 </h1>
<p> Fatal accidents happen the most in August (1.49%) and the least in June (1.19%). </p>

<h1> Question 12 </h1>
<h3> What is the most frequent accident severity? </h3>

In [21]:
severity = accident['Accident_Severity'].value_counts()
severity

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1> Insight 12 </h1>
<p> The majority of accidents are classified as slight, with a total of 563,801 occurrences. This is followed by serious accidents, which number 88,217, and fatal accidents, which are the least frequent, at 8,661. </p>

In [22]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
Year                                int32
Month                               int32
Day                                 int32
Day of the Week                     int32
Accident Year                       int32
Day_of_Week                        object
dtype: object

<h1> Question 13. </h1>
<h3> What is the most common light condition during accidents?</h3>

In [23]:
light_conditions = accident['Light_Conditions'].value_counts()
light_conditions

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h1> Insight 13 </h1>
<p> The highest number of accidents happened in daylight recorded with 484,880 incidents meanwhile129,335 occured in darkness </p>

<h1> Question 14. </h1>
<h3> What is the most frequent weather condition during accidents? </h3>

In [24]:
frequent_conditions = accident['Weather_Conditions'].value_counts()
frequent_conditions

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1> Insight 14 </h1>
<p> The most common weather conditions during accidents is "Fine no high winds" with 520,885 incidents.</p>

<h1> Question 15 </h1>
<p> What is the average number of vehicles involved in each accident? </p>

In [25]:
vehicle = accident['Number_of_Vehicles'].mean()
vehicle 

1.8312554205597575

<h1> Insight 14 </h1>
<p> The average number of vehicles involved in each accident is approximately 1.83.</p>

<h1> BiVariate </h1>
<h3> Question 15. </h3>
<h3> How does the weather affect the number of accidents? </h3>

In [26]:
no_of_accidents = accident.groupby('Weather_Conditions')['Accident Date'].count().sort_values(ascending=False)
no_of_accidents

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: Accident Date, dtype: int64

<h1> Insight 15 </h1>
<p> Rain, snow, fog, and high winds all contribute to a higher likelihood of accidents, with rain being the second most common weather condition for accidents. </p>

<h1> Question 16 </h1>
<h3> How do weather conditions, road surface conditions, and accident severity interact, and what impact do they have on the likelihood and severity of accidents?</h3>

In [27]:
road_severity = accident.groupby(['Weather_Conditions', 'Road_Surface_Conditions'])['Accident_Severity'].value_counts()
road_severity

Weather_Conditions  Road_Surface_Conditions    Accident_Severity
Fine + high winds   Dry                        Slight               4162
                                               Serious               758
                                               Fatal                 103
                    Flood over 3cm. deep       Slight                 14
                                               Fatal                   1
                                                                    ... 
unaccounted         Wet or damp                Serious               169
                                               Fatal                  29
                    unknown surface condition  Slight                550
                                               Serious                61
                                               Fatal                   1
Name: count, Length: 162, dtype: int64

<h1> Insight 16 </h1>
<p> Fine weather with high winds mostly result in slight injuries, but there are still some serious and fatal accidents</p>

<h1> Question 17 </h1>
<h3> Which road type experienced the most accidents? </h3>

In [28]:
road_type_accidents = accident['Road_Type'].value_counts()
road_type_accidents 

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h1> Insight 17 </h1>
<p> The road type that experienced the most accidents is the Single carriageway, with 492,143 accidents</p>

<h1> Question 18 </h1>
<h3> What is the monthly average of accidents?</h3>

In [29]:
accident_per_month = accident['Accident Date'].groupby(accident['Accident Date'].dt.month).count().mean()
accident_per_month 

55056.583333333336

<h1> Insight 18 </h1>
<p> The monthly average number of accidents is approximately 55,057. This means that, on average, there are about 55,057 accidents reported each month. </p>

<h1> Question 19. </h1>
<h3> What is the average number of accidents that occurred between 2019 and 2022?</h3>

In [30]:
accidents_in_2019 = accident[accident['Accident Date'].dt.year == 2019]['Accident Date'].count()
accidents_in_2022 = accident[accident['Accident Date'].dt.year == 2022]['Accident Date'].count()
total_accidents = accidents_in_2019 + accidents_in_2022
total_accidents

326534

<h1> Insight 19 </h1>
<p> The average number of accidents that occurred between 2019 and 2022 is 326,534. </p>

<h1> Question 20.</h1>
<h3> How many casualties were reported for each vehicle type in 2019? </h3>

In [31]:
casualties = df_2019.groupby(['Vehicle_Type'])['Number_of_Casualties'].count()
casualties

Vehicle_Type
Agricultural vehicle                        636
Bus or coach (17 or more pass seats)      10483
Car                                      130304
Data missing or out of range                  0
Goods 7.5 tonnes mgw and over              6247
Goods over 3.5t. and under 7.5t            2049
Minibus (8 - 16 passenger seats)            625
Motorcycle 125cc and under                 4313
Motorcycle 50cc and under                  2136
Motorcycle over 125cc and up to 500cc      2446
Motorcycle over 500cc                      7686
Other vehicle                              1588
Pedal cycle                                  68
Ridden horse                                  1
Taxi/Private hire car                      4230
Van / Goods 3.5 tonnes mgw or under        9303
Name: Number_of_Casualties, dtype: int64

<h1>Insight 20 </h1>
<p>Various vehicle types were involved in accidents during 2019, each contributing to a different number of casualties. Cars had the highest number of casualties, with a total of 130,304 reported. </p>

<h1> Question 21. </h1>
<h3> How are weather conditions connected to light conditions in road accidents?< <3h1>

In [32]:
weather_conditions = accident.groupby(['Weather_Conditions', 'Light_Conditions'])['Accident Date'].count()
weather_conditions

Weather_Conditions     Light_Conditions           
Fine + high winds      Darkness - lighting unknown        59
                       Darkness - lights lit            1852
                       Darkness - lights unlit            51
                       Darkness - no lighting            796
                       Daylight                         5796
Fine no high winds     Darkness - lighting unknown      3683
                       Darkness - lights lit           91951
                       Darkness - lights unlit          1700
                       Darkness - no lighting          24867
                       Daylight                       398684
Fog or mist            Darkness - lighting unknown        65
                       Darkness - lights lit             921
                       Darkness - lights unlit            37
                       Darkness - no lighting            862
                       Daylight                         1643
Other                  Darkness - 

<h1> Insight 21 </h1>
<p> Weather conditions like fog, snow, and rain are linked to more accidents in darkness, especially when there is no lighting or unlit lights. </p>

<h1> Question 22. </h1>
<h3> How do vehicle types relate to accident severity across different road types?</h3>

In [33]:
accident_severity = accident.groupby(['Vehicle_Type', 'Accident_Severity', 'Road_Type']).size()
accident_severity 


Vehicle_Type                         Accident_Severity  Road_Type         
Agricultural vehicle                 Fatal              Dual carriageway          6
                                                        One way street            1
                                                        Roundabout                0
                                                        Single carriageway       14
                                                        Slip road                 0
                                                                              ...  
Van / Goods 3.5 tonnes mgw or under  Slight             One way street          623
                                                        Roundabout             2081
                                                        Single carriageway    21532
                                                        Slip road               328
                                                        unaccounted             184
L

<h1> Insight 22 </h1>
<p> Agricultural vehicles tend to be involved in more fatal accidents on single carriageways, suggesting these vehicles may have a higher risk of severe accidents on certain road types. </p>